.travis.yml

```diff
@@ -20,25 +28,42 @@ env:
   global:
     - ANDROID_SDK_HOME=/opt/android/android-sdk
     - ANDROID_NDK_HOME=/opt/android/android-ndk
-  matrix:
-    - COMMAND='. venv/bin/activate && cd testapps/ && python setup_testapp_python3_sqlite_openssl.py apk --sdk-dir $ANDROID_SDK_HOME --ndk-dir $ANDROID_NDK_HOME --requirements libffi,sdl2,pyjnius,kivy,python3,openssl,requests,sqlite3,setuptools'
-    # overrides requirements to skip `peewee` pure python module, see:
-    # https://github.com/kivy/python-for-android/issues/1263#issuecomment-390421054
-    - COMMAND='. venv/bin/activate && cd testapps/ && python setup_testapp_python2_sqlite_openssl.py apk --sdk-dir $ANDROID_SDK_HOME --ndk-dir $ANDROID_NDK_HOME --requirements sdl2,pyjnius,kivy,python2,openssl,requests,sqlite3,setuptools'
-    - COMMAND='. venv/bin/activate && cd testapps/ && python setup_testapp_python2.py apk --sdk-dir $ANDROID_SDK_HOME --ndk-dir $ANDROID_NDK_HOME --bootstrap sdl2 --requirements python2,numpy'
-    # builds only the recipes that moved
-    - COMMAND='. venv/bin/activate && ./ci/rebuild_updated_recipes.py'
-
-before_script:
-  # we want to fail fast on tox errors without having to `docker build` first
-  - tox -- tests/ --ignore tests/test_pythonpackage.py
-  # (we ignore test_pythonpackage.py since these run way too long!! test_pythonpackage_basic.py will still be run.)
-
-script:
-  # Run a background process to make sure that travis will not kill our tests in
-  # case that the travis log doesn't produce any output for more than 10 minutes
-  - while sleep 540; do echo "==== Still running (travis, don't kill me) ===="; done &
-  - docker build --tag=p4a --file Dockerfile.py3 .
-  - docker run -e CI p4a /bin/sh -c "$COMMAND"
-  # kill the background process started before run docker
-  - kill %1
+
+jobs:
+  include:
+    - stage: lint
+      name: "Tox tests and coverage"
+      script:
+        # we want to fail fast on tox errors without having to `docker build` first
+        - tox -- tests/ --ignore tests/test_pythonpackage.py
+        # (we ignore test_pythonpackage.py since these run way too long!!
+        #  test_pythonpackage_basic.py will still be run.)
+      after_success:
+        - coveralls
+
+    - &testing
```

Comment:
I'm not aware of this syntax and also the `<<: *testing` below. Could you share me a doc to read or a keyword to look for?

Reply:
Oh yes, no problem, I learned this syntax for this pr, they are called `Yaml Anchors & Aliases` and are great to create templates, I discovered this in this article:
  - [Optimizing Mocha's Builds with Travis](https://boneskull.com/mocha-and-travis-ci-build-stages/#buildstage1smoke)

Also mentioned in [Travis docs examples](https://docs.travis-ci.com/user/build-stages/using-yaml-aliases/)
And here some more information: [atlassian - yaml anchors](https://confluence.atlassian.com/bitbucket/yaml-anchors-960154027.html)

Reply:
The `yaml Anchors & Aliases` that I wrote works the following way:

First we define a template named testing using the keywords `&testing`
```
    - &testing
      stage: test
      before_script:
        # build docker image
        - docker build --tag=p4a --file Dockerfile.py3 .
        # Run a background process to make sure that travis will not kill our tests in
        # case that the travis log doesn't produce any output for more than 10 minutes
        - while sleep 540; do echo "==== Still running (travis, don't kill me) ===="; done &
      script:
        - docker run -e CI -e TRAVIS_JOB_ID="$TRAVIS_JOB_ID" -e TRAVIS_BRANCH="$TRAVIS_BRANCH" p4a /bin/sh -c "$COMMAND"
      after_script:
        # kill the background process started before run docker
        - kill %1
```
The following code will take the above defined template, actually will also be part of the template but we override later for the other jobs:

```
      name: Python 3 basic
      # overrides requirements to skip `peewee` pure python module, see:
      # https://github.com/kivy/python-for-android/issues/1263#issuecomment-390421054
      env: COMMAND='. venv/bin/activate && cd testapps/ && python setup_testapp_python3_sqlite_openssl.py apk --sdk-dir $ANDROID_SDK_HOME --ndk-dir $ANDROID_NDK_HOME --requirements libffi,sdl2,pyjnius,kivy,python3,openssl,requests,sqlite3,setuptools'
```
For the following jobs we want to use the same scripts defined in  `&testing` so we use the keywords `<<:*testing` but we want to override the `name` and `env` defined in the template so we do the following:

```
    - <<: *testing
      name: Python 2 basic
      env: COMMAND='. venv/bin/activate && cd testapps/ && python setup_testapp_python2_sqlite_openssl.py apk --sdk-dir $ANDROID_SDK_HOME --ndk-dir $ANDROID_NDK_HOME --requirements sdl2,pyjnius,kivy,python2,openssl,requests,sqlite3,setuptools'
```

The same for the other jobs...

Reply:
Thanks for the explanation makes sense :+1: 
Our Travis file start to get more complex, but that's the way it is I guess.
Dropping python2 at some point may help slightly